src/StaticArraysCore.jl

```diff
@@ -44,21 +44,28 @@ const StaticVecOrMat{T} = Union{StaticVector{<:Any, T}, StaticMatrix{<:Any, <:An
 # The ::Tuple variants exist to make sure that anything that calls with a tuple
 # instead of a Tuple gets through to the constructor, so the user gets a nice
 # error message
-Base.@pure tuple_length(T::Type{<:Tuple}) = length(T.parameters)
-Base.@pure tuple_length(T::Tuple) = length(T)
-Base.@pure tuple_prod(T::Type{<:Tuple}) = length(T.parameters) == 0 ? 1 : *(T.parameters...)
-Base.@pure tuple_prod(T::Tuple) = prod(T)
-Base.@pure tuple_minimum(T::Type{<:Tuple}) = length(T.parameters) == 0 ? 0 : minimum(tuple(T.parameters...))
-Base.@pure tuple_minimum(T::Tuple) = minimum(T)
+tuple_length(T::Tuple) = length(T)
+tuple_prod(T::Tuple) = prod(T)
+tuple_minimum(T::Tuple) = minimum(T)
+
+tuple_svec(::Type{T}) where {T<:Tuple} = T.parameters
+
+# Julia `Base` provides a function for this purpose since v1.1: `fieldtypes`.
+# We don't use it yet, though, because it wrecks type inference with Julia
+# v1.6.
+Base.@pure tuple_tuple(::Type{T}) where {T<:Tuple} = (tuple_svec(T)...,)
+
+tuple_length(::Type{<:NTuple{n, Any}}) where {n} = n
+tuple_prod(::Type{T}) where {T<:Tuple} = mapreduce(Int, *, tuple_tuple(T), init = 1)
+tuple_minimum(::Type{Tuple{}}) = 0
+tuple_minimum(::Type{T}) where {T<:Tuple{Any,Vararg}} = mapreduce(Int, min, tuple_tuple(T), init = typemax(Int))
```

Comment:
This doesn't apply to empty tuples so `init` shouldn't be necessary I think?

Reply:
I'm a little worried that this more complex implementation could mess with inlining heuristics.

Reply:
> worried that this more complex implementation could mess with inlining heuristics

Doesn't seem like it. The output of the call is a constant, depending only on the parameter type, so the inlining cost is zero:
```julia-repl
julia> Base.print_statement_costs(stdout, StaticArraysCore.tuple_prod, (Type{Tuple{1,2,3,4,5,6,7,8,9,1,2,3}},))
tuple_prod(::Type{T}) where T<:Tuple @ StaticArraysCore ~/tmp/StaticArraysCore.jl/src/StaticArraysCore.jl:59
 0 1 ─     return 2177280
   

julia> Base.print_statement_costs(stdout, StaticArraysCore.tuple_minimum, (Type{Tuple{1,2,3,4,5,6,7,8,9,1,2,3}},))
tuple_minimum(::Type{T}) where T<:Tuple{Any, Vararg} @ StaticArraysCore ~/tmp/StaticArraysCore.jl/src/StaticArraysCore.jl:61
 0 1 ─     return 1
 ```

> This doesn't apply to empty tuples so `init` shouldn't be necessary I think?

Yeah, the keyword argument isn't necessary, however it's my understanding that it's a good practice to always provide the identity element to `reduce`-like calls. Although I guess it may not matter here, the call will get compiled to a constant integer either way.

Reply:
OK, thanks for checking!